src/Compilers/Core/Portable/StrongName/DesktopStrongNameProvider.cs

```diff
@@ -131,63 +133,22 @@ public DesktopStrongNameProvider(ImmutableArray<string> keyFileSearchPaths) : th
         /// </summary>
         /// <param name="tempPath">Path to use for any temporary file generation.</param>
         /// <param name="keyFileSearchPaths">An ordered set of fully qualified paths which are searched when locating a cryptographic key file.</param>
-        public DesktopStrongNameProvider(ImmutableArray<string> keyFileSearchPaths = default(ImmutableArray<string>), string tempPath = null)
+        public DesktopStrongNameProvider(ImmutableArray<string> keyFileSearchPaths = default(ImmutableArray<string>), string tempPath = null): this(keyFileSearchPaths, tempPath, null)
+        {
+        }
+
+        internal DesktopStrongNameProvider(ImmutableArray<string> keyFileSearchPaths, string tempPath, StrongNameFileSystem strongNameFileSystem)
         {
             if (!keyFileSearchPaths.IsDefault && keyFileSearchPaths.Any(path => !PathUtilities.IsAbsolute(path)))
             {
                 throw new ArgumentException(CodeAnalysisResources.AbsolutePathExpected, nameof(keyFileSearchPaths));
             }
 
+            FileSystem = strongNameFileSystem ?? new StrongNameFileSystem();
```

Comment:
`StrongNameFileSystem` is stateless. Consider creating a singleton instance.

Reply:
As we discussed, let's add a singleton instance for `StrongNameFileSystem` and make the `StrongNameFileSystem..ctor` `protected`. Then `DesktopStrongNameProvider.Equals` and `PortableStrongNameProvider.Equals` can compare `FileSystem` properties using `ReferenceEquals` and `StrongNameFileSystem` classes would not need to implement `Equals` or `GetHashCode`.